cache/class-vary-cache.php

```diff
@@ -41,36 +50,63 @@ class Vary_Cache {
 	 */
 	private static $cookie_expiry = MONTH_IN_SECONDS;
 
-	/** Nocache */
+	/**
+	 * Add "No-Cache" cookie for the user.
+	 *
+	 * This bypasses all requests from the VIP Cache.
+	 */
 	public static function set_no_cache_for_user() {
-		self::set_cookie( self::COOKIE_NO_CACHE, 1 );
+		if ( self::did_send_headers() ) {
+			return new WP_Error( 'did_send_headers', sprintf( 'Failed to register_groups (%s); cannot be called after the `send_headers` hook has fired.', implode( ', ', $groups ) ) );
+		}
+
+		if ( ! isset( $_COOKIE[ self::COOKIE_NO_CACHE ] ) ) {
+			self::set_cookie( self::COOKIE_NO_CACHE, 1 );
+		}
+
+		return true;
 	}
 
-	/** Clears the cache-busting flag */
+	/**
+	 * Clears the No-Cache cookie for the user.
+	 *
+	 * Restores caching behaviour for all future requests.
+	 */
 	public static function remove_no_cache_for_user() {
+		if ( self::did_send_headers() ) {
+			return new WP_Error( 'did_send_headers', sprintf( 'Failed to remove nocache cookie; cannot be called after the `send_headers` hook has fired.', implode( ', ', $groups ) ) );
```

Comment:
This probably needs to be `trigger_error` It's not firing if you call it in the `template_redirect` action. (Same as below)
Or we need to be clear that the developer needs to check for the error and deal with it accordingly.

Reply:
Ah, sorry, my testing instructions were flawed :)

We should follow the example of the `set_group` method here and return a WP_Error. We'll definitely want to update the usage docs.